core/src/main/java/org/apache/iceberg/BaseAddedRowsScanTask.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.util.List;
+import org.apache.iceberg.expressions.ResidualEvaluator;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+
+class BaseAddedRowsScanTask
+    extends BaseChangelogContentScanTask<AddedRowsScanTask, DataFile>
+    implements AddedRowsScanTask {
+
+  private final DeleteFile[] deletes;
+
+  BaseAddedRowsScanTask(int changeOrdinal, long commitSnapshotId, DataFile file, DeleteFile[] deletes,
+                        String schemaString, String specString, ResidualEvaluator residuals) {
+    super(changeOrdinal, commitSnapshotId, file, schemaString, specString, residuals);
+    this.deletes = deletes != null ? deletes : new DeleteFile[0];
+  }
+
+  @Override
+  protected AddedRowsScanTask self() {
+    return this;
+  }
+
+  @Override
+  public List<DeleteFile> deletes() {
+    return ImmutableList.copyOf(deletes);
+  }
+
+  @Override
+  protected Iterable<AddedRowsScanTask> splitUsingOffsets(List<Long> offsets) {
+    return () -> new OffsetsAwareSplitScanTaskIteratorImpl(this, offsets);
+  }
+
+  @Override
+  protected Iterable<AddedRowsScanTask> splitUsingFixedSize(long targetSplitSize) {
+    return () -> new FixedSizeSplitScanTaskIteratorImpl(this, targetSplitSize);
+  }
+
+  private static class SplitScanTaskImpl
+      extends SplitScanTask<SplitScanTaskImpl, AddedRowsScanTask, DataFile>
+      implements AddedRowsScanTask {
+
+    SplitScanTaskImpl(AddedRowsScanTask parentTask, long offset, long length) {
+      super(parentTask, offset, length);
+    }
+
+    @Override
+    public List<DeleteFile> deletes() {
+      return parentTask().deletes();
+    }
+
+    @Override
+    public SplitScanTaskImpl merge(ScanTask other) {
+      SplitScanTaskImpl that = (SplitScanTaskImpl) other;
+      return new SplitScanTaskImpl(parentTask(), start(), length() + that.length());
```

Comment:
Is it possible in  parent class to have abstract construction method:      

protected abstract ThisT newSplitScanTask(T parentTask, int start, int offset);

and then push merge() to parent, using this construction method?  Then we get merge() and canMerge() in the same class.

Reply:
I didn't quite follow the merge semantics here. we just add the length but lose track of `other` in the returned object.

Reply:
@szehon-ho, I am not sure that will be any cleaner as the only purpose of `merge` would be to directly call that new abstract method. I feel like it would pretty much the same.

@stevenzwu, that's the existing logic from `FileScanTask`. If two tasks are adjacent after split planning, we merge them together into one.

Reply:
Yea its not reducing a lot of code, just more symmetric for the reader.  The classes below (OffsetsAwareSplitScanTaskIteratorImpl and FixedSizeSplitScanTaskIteratorImpl) already have this 'newSplitTask' method, so we would just have one more just like that.

```
  private static class SplitScanTaskImpl
      extends SplitScanTask<SplitScanTaskImpl, AddedRowsScanTask, DataFile>
      implements AddedRowsScanTask {

    SplitScanTaskImpl(AddedRowsScanTask parentTask, long offset, long length) {
      super(parentTask, offset, length);
    }

    @Override
    public List<DeleteFile> deletes() {
      return parentTask().deletes();
    }

    @Override
    protected SplitScanTaskImpl newSplitTask(AddedRowsScanTask parentTask, long offset, long length) {
      return new SplitScanTaskImpl(parentTask, offset, length);
    }
  }
```

So the merge() is not just directly calling the newSplitTask but actually doing somewhat corresponding logic to canMerge(), which now is more clear in the same class.

```
  abstract static class SplitScanTask<ThisT, T extends ContentScanTask<F> & ChangelogScanTask, F extends ContentFile<F>>
      implements ContentScanTask<F>, ChangelogScanTask, MergeableScanTask<ThisT> {
   ...
       abstract ThisT newSplitTask(T parentTask, long offset, long length);
    
    @Override
    public boolean canMerge(ScanTask other) {
      if (getClass().equals(other.getClass())) {
        SplitScanTask<?, ?, ?> that = (SplitScanTask<?, ?, ?>) other;
        return changeOrdinal() == that.changeOrdinal() &&
            commitSnapshotId() == that.commitSnapshotId() &&
            file().equals(that.file()) &&
            start() + length() == that.start();

      } else {
        return false;
      }
    }

    @Override
    public ThisT merge(ScanTask other) {
      SplitScanTask that = (SplitScanTask) other;
      return newSplitTask(parentTask(), start(), length() + that.length());
    }
   ...
}
```

If we can somehow get rid of all these redundant classes with a lambda or something that would be the best, but probably its not possible.

Reply:
Okay, I've changed the approach to have `merge` in the parent. Let me know if that's what you meant.

Reply:
Yea, I didn't realize that you would remove the newSplitTask from the bottom two methods, leaving this one to be the only one.  But it's better to me.